data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/leader/ShardManager.java

```diff
@@ -148,22 +153,30 @@ private List<Shard> listShards(String streamArn, String lastEvaluatedShardId) {
         long startTime = System.currentTimeMillis();
         // Get all the shard IDs from the stream.
         List<Shard> shards = new ArrayList<>();
-        do {
-            DescribeStreamRequest req = DescribeStreamRequest.builder()
-                    .streamArn(streamArn)
-                    .limit(MAX_SHARD_COUNT)
-                    .exclusiveStartShardId(lastEvaluatedShardId)
-                    .build();
 
-            DescribeStreamResponse describeStreamResult = streamsClient.describeStream(req);
-            shards.addAll(describeStreamResult.streamDescription().shards());
+        try {
+            do {
+                DescribeStreamRequest req = DescribeStreamRequest.builder()
+                        .streamArn(streamArn)
+                        .limit(MAX_SHARD_COUNT)
+                        .exclusiveStartShardId(lastEvaluatedShardId)
+                        .build();
 
-            // If LastEvaluatedShardId is set,
-            // at least one more page of shard IDs to retrieve
-            lastEvaluatedShardId = describeStreamResult.streamDescription().lastEvaluatedShardId();
+                dynamoDBSourceAggregateMetrics.getStreamApiInvocations().increment();
+                DescribeStreamResponse describeStreamResult = streamsClient.describeStream(req);
+                shards.addAll(describeStreamResult.streamDescription().shards());
 
+                // If LastEvaluatedShardId is set,
+                // at least one more page of shard IDs to retrieve
+                lastEvaluatedShardId = describeStreamResult.streamDescription().lastEvaluatedShardId();
 
-        } while (lastEvaluatedShardId != null);
+
+            } while (lastEvaluatedShardId != null);
+        } catch(final InternalServerErrorException e) {
+            LOG.error("Received an internal server exception from DynamoDB while listing shards: {}", e.getMessage());
+            dynamoDBSourceAggregateMetrics.getStream5xxErrors().increment();
+            return shards;
```

Comment:
Should we rethrow the error here instead of potentially returning partial shards?

Reply:
I don't think there's anything wrong with returning partial shards, it just means we will only check those shards for child shards. The next time we run `describeStream` we will still go through the same shards again and create child shards for them (even the ones we return here) until 24 hours is up (when the shard would expire).